### PR TITLE
fix: Request rate limit log bug. Retrying all failures for s3. GITL-93

### PR DESCRIPTION
**Pull request by krackerman**

### CODE REVIEW

**Comment by krackerman** [comment]
Couple points of interest:
 - Now we are retrying everything.
 - We can create a create a `retryMap` or something similar to define retry limits with different error codes. 
 - S3 `backoffFunction, maxRetries` are not passed through on instantiation.

File: src/module/s3.js [hunk 1]
```diff
@@ -22,21 +20,12 @@ module.exports.S3 = ({
       await sleep(backoffFunction(count));
       try {
         // eslint-disable-next-line no-await-in-loop
-        return await call(action, opts, { expectedErrorCodes });
+        return await call(action, opts, {
+          expectedErrorCodes,
+          errorLogOverride: count < maxRetries
```
Comment (simlu) [inline src/module/s3.js:25]:
```suggestion
          logError: count < maxRetries
```

Comment (krackerman) [inline src/module/s3.js:25]: logError

File: test/module/s3.spec.js [hunk 2]
```diff
@@ -209,27 +211,21 @@ describe('Testing s3 Util', {
...
-      'Failed to submit s3:putObject\n'
-      + '{"errorCode":"SlowDown","action":"s3:putObject","opts":{"ContentType":"application/json",'
-      + '"ContentEncoding":"gzip","Bucket":"test-bucket-name","Key":"key","Body":{"type":"Buffer",'
-      + '"data":[31,139,8,0,0,0,0,0,2,3,171,86,74,73,44,73,84,178,130,80,181,0,185,30,67,221,15,0,0,0]}},'
-      + '"retryCount":0}'
-    ]);
+    expect(result).to.deep.equal({});
+    const errorLog = recorder.get();
+    const errorLogSplit = errorLog.map((e) => e.split('\n')[0]);
+    expect(errorLogSplit).to.deep.equal(['Request failed for s3.putObject()']);
```
Comment (krackerman) [inline test/module/s3.spec.js:228]: These logs are quite long. Is there a way you would want to see them?

File: test/module/s3.spec.js [hunk 3]
```diff
@@ -209,27 +211,21 @@ describe('Testing s3 Util', {
     expect(result).to.equal('2018-10-25T20:55:00.000Z/Collection Viewed.json.gz');
   });
 
-  it('Testing error rate exceeds retry count', async ({ capture, recorder }) => {
+  it('Testing error rate does not exceed retry count', async ({ recorder }) => {
     const s3 = S3({
       logger: console,
```
Comment (simlu) [inline test/module/s3.spec.js:216]: ?

File: test/module/s3.spec.js (hunk 2 above)
Comment (simlu) [reply test/module/s3.spec.js:228]: I'd like to see the retry count. How long are we talking? I'd just have them in here

File: test/module/s3.spec.js [hunk 4]
```diff
@@ -244,17 +240,31 @@ describe('Testing s3 Util', {
...
+      statusCode: 503
     });
-    expect(recorder.get()).to.deep.equal([
-      'Failed to submit s3:putObject\n'
-      + '{"errorCode":"SlowDown","action":"s3:putObject","opts":{"ContentType":"application/json",'
-      + '"ContentEncoding":"gzip","Bucket":"test-bucket-name","Key":"key","Body":{"type":"Buffer",'
-      + '"data":[31,139,8,0,0,0,0,0,2,3,171,86,74,73,44,73,84,178,130,80,181,0,185,30,67,221,15,0,0,0]}},'
-      + '"retryCount":0}'
+    const errorLog = recorder.get();
+    const errorLogSplit = errorLog.map((e) => e.split('\n')[0]);
```
Comment (simlu) [inline test/module/s3.spec.js:246]: lets have the full logs... or how long are they?

File: src/module/s3.js [hunk 5]
```diff
@@ -22,21 +21,15 @@ module.exports.S3 = ({
...
-          if (logger !== null) {
-            logger.warn(`Failed to submit ${action}\n${JSON.stringify({
-              errorCode: get(e, 'code'),
-              action,
-              opts,
-              retryCount: count
-            })}`);
-          }
-        } else {
+        if (get(e, 'code') !== 'SlowDown') {
```
Comment (simlu) [inline src/module/s3.js:29]:
```suggestion
        if (
        get(e, 'code') !== 'SlowDown'
        || count === maxRetries
        ) {
```
and remove lastError entirely?

File: test/module/s3.spec.js [hunk 6]
```diff
@@ -269,4 +224,63 @@ describe('Testing s3 Util', {
     expect(result).to.deep.equal({});
     expect(timeDiff).to.be.greaterThan(500);
   });
+
+  describe('Testing with logs', {
+    timestamp: 1607453027,
```
Comment (simlu) [inline test/module/s3.spec.js:229]: ...

**Review by simlu — APPROVED** [review]

**Pull request merged**